src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/AuthenticatorFactoryInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\DependencyInjection\Security\Factory;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.1
+ */
+interface AuthenticatorFactoryInterface
+{
+    /**
+     * Creates the authenticator service(s) for the provided configuration.
+     *
+     * @return string|string[] The authenticator service ID(s) to be used by the firewall
+     */
+    public function createAuthenticator(ContainerBuilder $container, string $id, array $config, string $userProviderId);
```

Comment:
We need to allow an array because the `CustomAuthenticatorFactory` allows users to specify multiple authenticator services. Do we want to *always* make this return an array so the interface has a return type"? Or do we not care?

Reply:
I would say "we do not care". I think the `CustomAuthenticatorFactory` will always be the only one requiring returning more than one authenticator. I didn't want to "mess up" the other method calls by returning a somewhat weird `[$authenticatorId]`. Anyway, I totally have no strong opinion on this matter :)